File: algorithms/java/单词拆分/单词拆分.java
// Source : https://leetcode.cn/problems/word-break/
// Author : Yanan Wang
// Date   : 2023-06-24

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> hash = new HashSet(wordDict);
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;
        s = " " + s;
        for (int i = 1; i <= n; i++) {
            for (int j = i; j > 0; j--) {
                if (dp[j - 1] && hash.contains(s.substring(j, i + 1))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
}

File: algorithms/python/单词拆分/单词拆分.py
# Source : https://leetcode.cn/problems/word-break/description/
# Author : Yanan Wang
# Date   : 2024-01-17

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        for i in range(1, n + 1):
            for j in range(i, 0, -1):
                if dp[j - 1] and (s[j-1:i] in wordDict):
                    dp[i] = True
        
        return dp[n]